clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -3181,3 +3189,80 @@ LangAS CIRGenModule::getGlobalVarAddressSpace(const VarDecl *D) {
 
   return getTargetCIRGenInfo().getGlobalVarAddressSpace(*this, D);
 }
+
+mlir::StringAttr CIRGenModule::buildAnnotationString(StringRef Str) {
+  auto &Astr = AnnotationStrings[Str];
+  if (Astr)
+    return Astr;
+  Astr = builder.getStringAttr(Str);
+  return Astr;
+}
+
+mlir::ArrayAttr CIRGenModule::buildAnnotationArgs(const AnnotateAttr *Attr) {
+  ArrayRef<Expr *> Exprs = {Attr->args_begin(), Attr->args_size()};
+  if (Exprs.empty()) {
+    return mlir::ArrayAttr::get(builder.getContext(), {});
+  }
+  llvm::FoldingSetNodeID ID;
+  for (Expr *E : Exprs) {
+    ID.Add(cast<clang::ConstantExpr>(E)->getAPValueResult());
+  }
+  auto &Lookup = AnnotationArgs[ID.ComputeHash()];
+  if (Lookup)
+    return Lookup;
+
+  llvm::SmallVector<mlir::Attribute, 4> Args;
+  Args.reserve(Exprs.size());
+  for (Expr *E : Exprs) {
+    if (const auto StrE =
+            ::clang::dyn_cast<clang::StringLiteral>(E->IgnoreParenCasts())) {
+      // Add trailing null character as StringLiteral->getString() does not
+      Args.push_back(
+          buildAnnotationString(std::string(StrE->getString()) + '\0'));
+    } else if (const auto IntE = ::clang::dyn_cast<clang::IntegerLiteral>(
+                   E->IgnoreParenCasts())) {
+      Args.push_back(mlir::IntegerAttr::get(
+          mlir::IntegerType::get(builder.getContext(),
+                                 IntE->getValue().getBitWidth()),
+          IntE->getValue()));
+    } else {
+      llvm_unreachable("NYI");
+    }
+  }
+  // Create const struct attr for these arguments
+  mlir::ArrayAttr ArgsArrayAttr = builder.getArrayAttr(Args);
+  Lookup = ArgsArrayAttr;
```

Comment:
I already mentioned this in the past round: please use camelBack for coding guidelines, see https://llvm.github.io/clangir/GettingStarted/coding-guideline.html for more information.

All variable names and functions should start lowercase, please let's not go over that every PR and every round of review.

Reply:
Glad to hear that and you clarify it. The fact is that lots of code in this file are NOT using camelBack probably to look similar to OG code. Therefore, it was hard to decide what convention to follow as this code has OG counterpart too. 
But in future, I'll probably use camelBack for code added into this file. unless explicitly agreed on using OG style.